src/flask_pyoidc/message_factory.py

```diff
@@ -0,0 +1,6 @@
+from oic.oauth2.message import AccessTokenResponse, CCAccessTokenRequest, MessageTuple, OauthMessageFactory
+
+
+class CCMessageFactory(OauthMessageFactory):
```

Comment:
What is this needed for?
I see the usage of it, but don't understand why/how it's needed 🙂

Reply:
Pyoidc message factories handles request and response classes to parse data respectively. But request and response format vary with the Oauth2.0 flow used. In order to use appropriate classes for the type of data that you are about to send and receive, message factories return the classes that will consume the data. E.g. `oic.oauth2.message.OauthMessageFactory` exists by default.

```python
class OauthMessageFactory(MessageFactory):
    """Factory that knows Oauth2.0 message types."""

    authorization_endpoint = MessageTuple(AuthorizationRequest, AuthorizationResponse)
    token_endpoint = MessageTuple(AccessTokenRequest, AccessTokenResponse)
    refresh_endpoint = MessageTuple(RefreshAccessTokenRequest, AccessTokenResponse)
    # Message as a placeholder ...
    resource_endpoint = MessageTuple(ResourceRequest, Message)
    configuration_endpoint = MessageTuple(Message, ASConfigurationResponse)
```

If you are about to make authorization code flow access token request to the `token_endpoint` of the IdP, `OauthMessageFactory.token_endpoint` will return `AccessTokenRequest` which will construct your request and `AccessTokenResponse` which will parse the response.

But for client credentials flow, the request format is different. So I've created `CCMessageFactory` that overrides `token_endpoint` to handle request and response appropriately. This flexibility allows you to implement custom classes for these formats which then can be passed as a message factory to Pyoidc.

I found this solution in one of their test cases:

https://github.com/OpenIDC/pyoidc/blob/0d316fc0baca21b16164ef17fb275979d178f4b6/tests/test_oauth2.py#L568-L586

Reply:
Makes sense, thanks for clarifying! 👍 
(I expected `AccessTokenRequest` to be a generic superclass of possible token request, but I see now it is not - it is only for the "Authorization Code Flow".)